### PR TITLE
[RFC] Add TextYankPost and TextDeletePost autocmds

### PR DESCRIPTION
**Pull request by Shougo**
I have ported from https://github.com/Silex/vim/commit/de53ab72c89affa8ba77536ed8920751c037d127

It improves yankring and neoyank behavior.

I have added the simple tests.

### CODE REVIEW

File: test/functional/autocmd/text_deletepost.lua [hunk 1]
```diff
@@ -0,0 +1,25 @@
+local helpers = require('test.functional.helpers')
+local clear, eval, eq, neq = helpers.clear, helpers.eval, helpers.eq, helpers.neq
```
Comment (watiko) [inline test/functional/autocmd/text_deletepost.lua]: Some functions are missing: `execute`, `feed`, `nvim`.

File: test/functional/autocmd/text_deletepost.lua [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+describe('TextDeletePost', function()
+    before_each(function()
```
Comment (watiko) [inline test/functional/autocmd/text_deletepost.lua]: Normally 2 space indentation is better.

File: test/functional/autocmd/text_deletepost.lua [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        clear()
+    end)
+
+    describe('au TextDeletePost', function()
+        it('is executed after delete', function()
+            feed('ifoo<ESC>')
+            execute('let g:foo = 0')
+            nvim('command', 'au! TextDeletePost * let g:foo = 1')
```
Comment (watiko) [inline test/functional/autocmd/text_deletepost.lua]: In this case `execute('au! TextDeletePost * let g:foo = 1')` is enough.

File: src/nvim/ops.c [hunk 4]
```diff
@@ -1585,6 +1586,10 @@ int op_delete(oparg_T *oap)
 
   msgmore(curbuf->b_ml.ml_line_count - old_lcount);
 
+  ++textlock;
```
Comment (watiko) [inline src/nvim/ops.c]: In this case I would suggest replaces `pre{increment,decrement}` which `post{increment,decrement}` even if this is just porting.

see: https://neovim.io/develop/style-guide.xml#Postincrement_and_Postdecrement

**Comment by Shougo** [comment]
Thank you!
Fixed.

**Comment by bfredl** [comment]
Very nice, I will try it :+1:

**Comment by bfredl** [comment]
I think is slight potential performance/semantic problem. Consider the user sets `clipboard=unnamed` and does a yank. Now the clipboard manager needs to to `[getreg(v:register), getregtype(v:register)]` which amounts to two pastes.It is very unlikely but possible that the system clipboard changes in this time, at least there is two extra roundrips (except on X11 where we circumvent this problem). I think it would be useful if the autocmd sets say `v:event.regcontents` and `v:event.regtype` and maybe even `v:event.regname` which contains the actual register name (including the empty string for a bare `yy`) unlike `v:register` which contains the adjusted register and cannot distiguish `""yy` and `yy` 

And while were are bikeshedding, we could consider just having one autocommand and having a event field that says "yank", "delete" and maybe even "change" (or possibly just the op name, as I think  `oap` propagates it).

**Comment by Shougo** [comment]
@bfredl I think it is better.  But it is too complex.

**Comment by bfredl** [comment]
Well, the importart part was `v:event.regcontents` and `v:event.regtype`, the pieces needed should already be there. But if you think is complex I'll be happy to look into it myself.

**Comment by Shougo** [comment]
>  But if you think is complex I'll be happy to look into it myself.

I think it is better.  Thanks.

**Comment by bfredl** [comment]
I implemented my proposal [here](https://github.com/bfredl/neovim/tree/yank). It needs cleanup and doc/test update, but one could try it using:

au! TextYankPost * echo v:event

**Comment by Shougo** [comment]
> I implemented my proposal here. It needs cleanup and doc/test update, but one could try it using:

Thank you!

**Comment by bfredl** [comment]
Ok, superseded by #4304

**Pull request closed without merging**